src/app/modules/app/app.js

```diff
@@ -123,18 +139,106 @@
+                if(languages.length <= numberOfTopLanguages) {
+                    $scope.topLanguages = languages.sort(compareLang);
+                    $scope.otherLanguages = [];
+                }
+                else {
+                    //Split the list of languages into top and other, and sort the lists alphabetically
+                    $scope.topLanguages = languages.slice(0, numberOfTopLanguages).sort(compareLang);
+                    $scope.otherLanguages = languages.slice(numberOfTopLanguages, languages.length).sort(compareLang);
+                }
+
+                for(var i = 0; i < $scope.otherLanguages.length; i++) {
+                    if($scope.otherLanguages[i].isoCode == $scope.currentLang) {
+                        $scope.topLanguages.unshift($scope.otherLanguages.splice(i, 1)[0]);
+                        $scope.topLanguages.sort(compareLang);
+                        i = $scope.otherLanguages.length;
+                    }
+                }
             });
-            $scope.topTags = tagService.getTags();
+
+                //This is the global list of all categories of tags
+                //The tags are separated out by these categories on the sidebar of the browse view
+                //id is the usage in the tag (e.g. "region.MyRegion") and displayName is the header
+                $scope.tagCategories = [
+                    {id: 'topic', displayName: 'Topics'},
+                    {id: 'region', displayName: 'Regions'},
+                    {id: 'level', displayName: 'Reading Levels'}
+                ];
+
+                //This is the object which will hold all of the tag names
+                $scope.tags = {topic: {top: [], other: []}};
+
+                tagService.getTags().then(function(tags) {
+                    //Get the names out of the tags; we don't care about the other properties
+                    var tagNames = tags.map(function(item) {
+                        return item.name;
+                    });
+
+                    //This is the number of tags that will be shown in each category of tag before the (n more...)
+                    var numberOfTopTags = 100;
+
+                    var i, j, cat;
+                    var categoryRegex = {};
+
+                    //Set up objects and regexes to save processing time
+                    for(i = 0; i < $scope.tagCategories.length; i++) {
+                        cat = $scope.tagCategories[i].id;
+                        $scope.tags[cat] = {top: [], other: []};
+                        categoryRegex[cat] = new RegExp('^' + cat + '\\.');
+                    }
+
+                    //Loop through tags
+                    for(i = 0; i < tagNames.length; i++) {
+                        var isPartOfCategory = false;
+
+                        //Check if tag belongs to a category
+                        for(j = 0; j < $scope.tagCategories.length; j++) {
```

Comment:
Is the function still too long? I am reluctant to move stuff into helper functions only being used once because they will occupy space in some scope or other. If this still needs to be shorter, what scope do I put the helper functions in?